weave/workers/weave_scorer.py

```diff
@@ -0,0 +1,445 @@
+        end=tsi.EndedCallSchemaForInsert(
+            project_id=project_id,
+            id=call_start_res.id,
+            ended_at=datetime.now(),
+            output=result,
+            summary={},
+        )
+    )
+
+    server.call_end(call_end_req)
+
+    return call_start_res.id, result
+
+
+def _get_score_call(score_call_id: str, project_id: str) -> Call:
+    """Gets a score call from the DB."""
+    server = get_trace_server()
+
+    call_req = tsi.CallsQueryReq(
+        project_id=project_id, filter=tsi.CallsFilter(call_ids=[score_call_id])
+    )
+
+    calls = server.calls_query(call_req).calls
+
+    return build_client_call(calls[0])
+
+
+async def apply_scorer(
+    monitor_internal_ref: InternalObjectRef,
+    scorer: Scorer,
+    call: Call,
+    project_id: str,
+    wb_user_id: str,
+) -> None:
+    """Actually apply the scorer to the call."""
+    score_call_id, result = _do_score_call(scorer, call, project_id)
+
+    score_call = _get_score_call(score_call_id, project_id)
+
+    call_ref = InternalCallRef(project_id=project_id, id=call.id)  # type: ignore
+    score_call_ref = InternalCallRef(project_id=project_id, id=score_call.id)  # type: ignore
+
+    results_json = to_json(result, project_id, None)  # type: ignore
+    payload = {"output": results_json}
+
+    server = get_trace_server()
+
+    feedback_req = FeedbackCreateReq(
+        project_id=project_id,
+        weave_ref=call_ref.uri(),
+        feedback_type=RUNNABLE_FEEDBACK_TYPE_PREFIX + "." + scorer.name,  # type: ignore
+        payload=payload,
+        runnable_ref=scorer.__dict__["internal_ref"].uri(),
+        call_ref=score_call_ref.uri(),
+        wb_user_id=wb_user_id,
+        trigger_ref=monitor_internal_ref.uri(),
+    )
+
+    server.feedback_create(feedback_req)
+
+
+async def process_ended_call(ended_call: tsi.EndedCallSchemaForInsert) -> None:
+    project_id = ended_call.project_id
+
+    active_monitors_ref_user_ids = get_active_monitors(project_id)
+
+    for monitor, monitor_internal_ref, wb_user_id in active_monitors_ref_user_ids:
+        await process_monitor(monitor, monitor_internal_ref, ended_call, wb_user_id)  # type: ignore
+
+
+def _call_processor_done_callback(
+    task: asyncio.Task,
+    msg: Message,
+    consumer: KafkaConsumer,
+    call_processors: set[asyncio.Task],
+) -> None:
+    call_processors.discard(task)
+
+    try:
+        task.result()
+        consumer.commit(msg)
+    except Exception as e:
+        logger.exception(
+            f"Error processing message: {e.__class__.__name__} {e}", exc_info=e
+        )
+
+
+async def process_kafka_message(
+    msg: Message, consumer: KafkaConsumer, call_processors: set[asyncio.Task]
+) -> bool:
+    """Process a single Kafka message and create a task for it."""
+    try:
+        ended_call = tsi.EndedCallSchemaForInsert.model_validate_json(
+            msg.value().decode("utf-8")
+        )
+        logger.info("Processing ended call %s", ended_call.id)
+
```

Comment:
there is a critical item here: we cannot be sure that the start event is indeed inserted before the end event. There are a few ways to handle this. One simple method could be to re-queue the item with some backoff bookkeeping

Reply:
the way I had proposed during scoping is actually to have another queue (completed calls) that contains calls that are fully matched up. This would allow for other downstream applications to operate on already-matched calls.

Reply:
Do we need to care that the start event was inserted?
All we need is the end event, which contains the Op output. No?